src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,143 +103,211 @@ public class FromBase64Transform : ICryptoTransform
 
         public int TransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount != InputBlockSize is not allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount, InputBlockSize);
+            if (outputBuffer == null) ThrowHelper.ThrowArgumentNull(ThrowHelper.ExceptionArgument.outputBuffer);
+
+            if (_inputBuffer == null)
+            {
+                ThrowHelper.ThrowObjectDisposed();
+            }
+
+            Debug.Assert(InputBlockSize == 4);
+            Span<byte> tmpBuffer = stackalloc byte[4];
 
-            int effectiveCount;
-            byte[] temp = GetTempBuffer(inputBuffer, inputOffset, inputCount, out effectiveCount);
+            tmpBuffer = GetTempBuffer(inputBuffer.AsSpan(inputOffset, inputCount), tmpBuffer);
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
 
-            if (effectiveCount + _inputIndex < 4)
+            // To less data to decode: save data to _inputBuffer, so it can be transformed later
+            if (bytesToTransform < InputBlockSize)
             {
-                Buffer.BlockCopy(temp, 0, _inputBuffer, _inputIndex, effectiveCount);
-                _inputIndex += effectiveCount;
+                tmpBuffer.CopyTo(_inputBuffer.AsSpan(_inputIndex));
+
+                _inputIndex = bytesToTransform;
                 return 0;
             }
 
-            byte[] result = ConvertFromBase64(temp, effectiveCount);
+            ConvertFromBase64(tmpBuffer, outputBuffer.AsSpan(outputOffset), out int consumed, out int written);
+            Debug.Assert(consumed == bytesToTransform);
 
-            Buffer.BlockCopy(result, 0, outputBuffer, outputOffset, result.Length);
-
-            return result.Length;
+            return written;
         }
 
         public byte[] TransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount > InputBlockSize is allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+
+            if (_inputBuffer == null)
+            {
+                ThrowHelper.ThrowObjectDisposed();
+            }
+
+            if (inputCount == 0)
+            {
+                return Array.Empty<byte>();
+            }
+
+            // The common case is <= 4
+            byte[] tmpBufferArray = null;
+            Span<byte> tmpBuffer = stackalloc byte[4];
+            if (inputCount > 4)
+            {
+                tmpBuffer = tmpBufferArray = ArrayPool<byte>.Shared.Rent(inputCount);
```

Comment:
Since this is in a crypto namespace, and the rented buffer is not being passed out to a callback or unsealed virtual dispatch, `CryptoPool` should be used, vs (directly) using ArrayPool.Shared.

Reply:
Thansk for the hint. I'll update the pool to the `CryptoPool`.